api/src/main/java/me/lokka30/treasury/api/economy/currency/CurrencyManager.java

```diff
@@ -0,0 +1,248 @@
+package me.lokka30.treasury.api.economy.currency;
+
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.UUID;
+import java.util.concurrent.ConcurrentHashMap;
+import org.jetbrains.annotations.NotNull;
+
+/**
+ * Represents a manager of {@link Currency currencies}.
+ *
+ * @author MrIvanPlays
+ * @since v1.0.0
+ */
+public enum CurrencyManager {
+    INSTANCE;
```

Comment:
Shoudln't this be `CurrencyManager INSTANCE;` ?

Reply:
I decided to use enum cuz that's the only way you can _really_ enforce 1 instance of a file. To answer your question, no.

Reply:
Seems kinda hacky. 

```java
public static final CurrencyManager INSTANCE = new CurrencyManager();

private CurrencyManager() {
  if(INSTANCE != null) throw new IllegalCallerException("Tryed to create more than one instances of CurrencyManager");
}

```
Doesn't this work?

Reply:
you can't have anything in the constructor. this code at the point of the constructor call `INSTANCE` is already non null, leading to the exception. 

enforcing a single instance is something full of compromises, and as I said the only way to _really_ enforce a single instance is by using an enum. if we use another method, someone with bad intentions can break all the currencies and leave permanent damage on the economy.

Reply:
```java
public class HelloWorld{

     public static void main(String []args){
        Singleton.INSTANCE.sayHi();
     }
     
     static class Singleton {
        public static final Singleton INSTANCE = new Singleton();
  
          private Singleton() {
            if(INSTANCE != null)
        	  throw new RuntimeException("Tryed to create more than one instances of Singleton");
          }
          
          public void sayHi() {
        	System.out.println("Hi!");
          }
  
    }
     
}
```
This runs without problems, I think I'm not understanding you

Reply:
Using a single element enum isn't super common practice but there's nothing inherently wrong with it compared to more traditional singleton patterns. Theoretically it's roughly the equivalent of any other similar security measures - it effectively is a final class with a public final field pre-instantiated. A sufficiently determined person could still replace it, but they'd have to really be trying. It just saves on writing boilerplate code to create the instance.

Reply:
> A sufficiently determined person could still replace it, but they'd have to really be trying

to add to this, consider we have this code: 
```java
public class DoesntMatter {

  private static DoesntMatter instance;

  public static DoesntMatter getInstance() {
    if (instance == null) {
      instance = new DoesntMatter();
    }
    return instance;
  }

  private DoesntMatter() {
  }

  // whatever

}
```

the `instance` field can easily be altered via reflection:
```java
Field field = DoesntMatter.class.getDeclaredField("instance");
field.setAccessible(true);
Constructor constructor = DoesntMatter.class.getDeclaredConstructor();
field.set(null, constructor.newInstance());
```

there I just altered the instance of that class ^, potentially clearing all the maps and all the stuff this class needs to function properly during runtime.

whilst as Jikoo said with an enum they should really be trying, but with such designs - 4 lines and I'm able to ruin the entire plugin (make it spit errors, do whatever).

Reply:
I see